<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Understanding R</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/lumen.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="site_libs/highlight/default.css"
      type="text/css" />
<script src="site_libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 54px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 59px;
  margin-top: -59px;
}

.section h2 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h3 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h4 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h5 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h6 {
  padding-top: 59px;
  margin-top: -59px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.9em;
  padding-left: 5px;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
  padding-left: 10px;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">UCVM R Bootcamp</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Course Materials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="2_understanding_R.html">Understanding R</a>
    </li>
    <li>
      <a href="3_prac_data_management_1.html">Practical data management 1</a>
    </li>
    <li>
      <a href="4_prac_data_management_2.html">Practical data management 2</a>
    </li>
    <li>
      <a href="5_data_viz.html">Data visualization</a>
    </li>
    <li>
      <a href="6_intro_to_bioconductor.html">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="7_wrapup.html">Reproducability and wrapup</a>
    </li>
  </ul>
</li>
<li>
  <a href="references.html">
    <span class="fa fa-external-link"></span>
     
    References
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    About
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="about.html">
        <span class="fa fa-info-circle"></span>
         
        Who we are
      </a>
    </li>
    <li>
      <a href="license-summary.html">
        <span class="fa fa-creative-commons"></span>
         
        License
      </a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ucvm/rbootcamp">
    <span class="fa fa-github fa-2x"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Understanding R</h1>

</div>


<p>Outline</p>
<ul>
<li>R History</li>
<li>Programming concepts</li>
<li>Environments</li>
<li>How to work with objects: subsetting, indexing, apply functions</li>
<li>Other important key R functions including basic statistics</li>
</ul>
<div id="r-history" class="section level2">
<h2>R History</h2>
<p>What is R?</p>
<p>R is calculator</p>
<pre><code>2 + 2</code></pre>
<p>NO!! R is a programming languange. Specifically it’s a programming language built for statistics. And that’s what it’s best at.</p>
<p>R is a dialect of the S languange. S was developed by John Chambers at Bell Labs in 1976 and still exists today although hasn’t changed much since 1998. The philosopy behind S (and R) was to allow users to begin in an interactive enviroment that didn’t explicitly feel like programming. Then, as their needs and skills grew they could move into more of the programming aspects. This helps us understand some of why R is the way it is.</p>
<p>R began life in New Zealand, developed by Ross Ihaka and Robert Gentleman in 1991. It was made available to the public in 1993 and in 1995 R was licensed with the GNU General Public License, making it free and open-source. Version 1.0.0 was released in 2000 and the most recent version, 3.3.1 was released on June 21.</p>
<div id="some-key-features-of-r" class="section level3">
<h3>Some key features of R</h3>
<ul>
<li>R runs on almost all platforms and operating systems.<br />
</li>
<li>It’s free</li>
<li>The core is quite lean - most funcionality is found in modular packages.</li>
<li>Very powerful graphics and statistics capabilities</li>
<li>Actively developed and a very active user community</li>
<li>Rich and robust package repository (CRAN and Bioconductor)</li>
<li>Excellent interactive capabilities - good for rapid development and data analysis</li>
</ul>
</div>
</div>
<div id="packages" class="section level2">
<h2>Packages</h2>
<p>Packages are simply bits of code, external to the core R code that are designed to perform a specific function. The vast majority of the usefulness and functionality of R resides in packages. These packages live in online repositories and can be installed on your own system to be used.</p>
<p>R has a well defined system of packages, requiring package authors to document well and test installation thouroughly. This means that <em>most</em> packages will install easily on any system. For most R packages the central repository is CRAN (The Comprehensive R Archive Network), however, most bioinformatic packages live in another repository called Biocondcutor. Dispite differences in content and appearances, these essentially function in the same way.</p>
<div id="installing-packages" class="section level3">
<h3>Installing packages</h3>
<p>Packages need only be installed once, although you may have to reinstall when upgrading R or when you want to use a newer version of a package.</p>
<p>To install from CRAN all one needs to do is:</p>
<pre><code>install.packages(&quot;dplyr&quot;)</code></pre>
<p>If you’re not using RStudio then you may be asked to select a mirror. Just choose the location geographically closest to you.</p>
<p>Bioconductor is slightly different - we’ll cover that in more detail in a later session.</p>
</div>
<div id="using-packages" class="section level3">
<h3>Using packages</h3>
<p>Once installed all the functions in a package are available to be used.</p>
<pre class="r"><code>dplyr::glimpse(iris)
## Observations: 150
## Variables: 5
## $ Sepal.Length &lt;dbl&gt; 5.1, 4.9, 4.7, 4.6, 5.0, 5.4, 4.6, 5.0, 4.4, 4.9,...
## $ Sepal.Width  &lt;dbl&gt; 3.5, 3.0, 3.2, 3.1, 3.6, 3.9, 3.4, 3.4, 2.9, 3.1,...
## $ Petal.Length &lt;dbl&gt; 1.4, 1.4, 1.3, 1.5, 1.4, 1.7, 1.4, 1.5, 1.4, 1.5,...
## $ Petal.Width  &lt;dbl&gt; 0.2, 0.2, 0.2, 0.2, 0.2, 0.4, 0.3, 0.2, 0.2, 0.1,...
## $ Species      &lt;fctr&gt; setosa, setosa, setosa, setosa, setosa, setosa, ...</code></pre>
<p>Here the name of the package is provide followed by two colons and then the name of the function you want to use. The <code>::</code> loads the package into memory and allows you to access all of the functions.</p>
<p>However, this can get tedious typing out the package name everytime so R provides a function <code>library()</code> used to attach packages. The library function first loads then “attaches” the package. Basically this means you can now use functions from a package without typing the package names. Packages are attached in your current session and need to be attached every time you start a new session. Technically what is happening here is that when attaching a package R puts those functions in your search path, the place R looks first for objects and functions.</p>
<pre class="r"><code>search()
##  [1] &quot;.GlobalEnv&quot;        &quot;package:knitr&quot;     &quot;package:stats&quot;    
##  [4] &quot;package:graphics&quot;  &quot;package:grDevices&quot; &quot;package:utils&quot;    
##  [7] &quot;package:datasets&quot;  &quot;package:methods&quot;   &quot;Autoloads&quot;        
## [10] &quot;package:base&quot;
library(&quot;dplyr&quot;)
## 
## Attaching package: &#39;dplyr&#39;
## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag
## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union
search()
##  [1] &quot;.GlobalEnv&quot;        &quot;package:dplyr&quot;     &quot;package:knitr&quot;    
##  [4] &quot;package:stats&quot;     &quot;package:graphics&quot;  &quot;package:grDevices&quot;
##  [7] &quot;package:utils&quot;     &quot;package:datasets&quot;  &quot;package:methods&quot;  
## [10] &quot;Autoloads&quot;         &quot;package:base&quot;

glimpse(iris)
## Observations: 150
## Variables: 5
## $ Sepal.Length &lt;dbl&gt; 5.1, 4.9, 4.7, 4.6, 5.0, 5.4, 4.6, 5.0, 4.4, 4.9,...
## $ Sepal.Width  &lt;dbl&gt; 3.5, 3.0, 3.2, 3.1, 3.6, 3.9, 3.4, 3.4, 2.9, 3.1,...
## $ Petal.Length &lt;dbl&gt; 1.4, 1.4, 1.3, 1.5, 1.4, 1.7, 1.4, 1.5, 1.4, 1.5,...
## $ Petal.Width  &lt;dbl&gt; 0.2, 0.2, 0.2, 0.2, 0.2, 0.4, 0.3, 0.2, 0.2, 0.1,...
## $ Species      &lt;fctr&gt; setosa, setosa, setosa, setosa, setosa, setosa, ...</code></pre>
<p>There is some confusion on why they are called packages but you used the <code>library()</code> function to attach them. The correct terminology here is that individual packages are stored in your ‘library’. You use the <code>library()</code> function to load and attach a package <em>from</em> your library.</p>
</div>
<div id="namespaces" class="section level3">
<h3>Namespaces</h3>
<p>An important concept to be aware of when using packages is namespaces. Given the thousands of packages available it is quite likely that function names will overlap. If two functions have the same name and both are attached, R will by default use the one attached most recently. You can see this by looking at the order of the packages in the search path <code>search()</code> function.</p>
<p>To avoid problems and bizzare errors you can specify which function to use by using the <code>::</code> notation as above to explicity indicate which function you’d like to use.</p>
<p>Pro tip: If your getting strange errors from a function that previously worked fine try typing <code>?function_name</code>. In RStudio if there are multiple functions attached with the same name then the help window will give you links for both functions and the one at the top of the list is the one R uses by default.</p>
</div>
</div>
<div id="programming-concepts" class="section level2">
<h2>Programming concepts</h2>
<p>R is a programming language and in order to use it effectivly it is important to understand some basic programming concepts.</p>
<div id="variables-objects" class="section level3">
<h3>Variables / Objects</h3>
<p>Think of a variable (also called objects) as container to hold a value. Containers are defined to hold specific types of values and we can give the informative names. These are the data types that we talked about in the last lesson. They can also be abitrarily complex. For example a DESeq object, used in analzying RNA-seq data contains the expression data, additional annotations on the genes, sample data, experiment details and results from any statistics run.</p>
<p>Objects are used to structure data and reduce repetition. They are passed to functions and can be modified.</p>
<pre class="r"><code>x = 2
y = c(1,2,3)

x + y
## [1] 3 4 5

sum(x,y)
## [1] 8</code></pre>
</div>
<div id="functions" class="section level3">
<h3>Functions</h3>
<p>A function is a small bit of code designed to perform a single task over and over. Given a set of inputs a well designed function takes those, performs a pre-defined, single task and returns some type of output.</p>
<p>Functions are first-class citizens in R. They are used extensively, can be passed as parameters to other functions and functions can return other functions as output.</p>
<p>Most functions you use will be either base R functions (functions that are a part of the core R language) or functions defined by external packages. However, defining your own functions in R is very easy and can be very helpful to reduce repetition and mistakes in your analysis.</p>
<pre class="r"><code># base functions
seq(1:5)
## [1] 1 2 3 4 5
sum(c(2, 39, 22))
## [1] 63

# parameters are given using a single &#39;=&#39;
x = rnorm(n = 100, mean = 35, sd = 2)

# but they don&#39;t have to be named as long as you give them in the right order
y = rnorm(100, 46, 1.5)
plot(x, y)</code></pre>
<p><img src="2_understanding_R_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<pre class="r"><code>
# functions can be nested
print(paste(&quot;The mean of the x vector is&quot;, round(mean(x))))
## [1] &quot;The mean of the x vector is 35&quot;

# a simple function of our own devising
say_hello = function(name) {
    say = paste(&quot;Hello&quot;, name)
    print(say)
}

say_hello(&quot;Matthew&quot;)
## [1] &quot;Hello Matthew&quot;</code></pre>
</div>
<div id="conditionals" class="section level3">
<h3>Conditionals</h3>
<p>Programming is often about making choices; if this then that. This is how we can control the program flow based on things like user input, features of our data, or the output of another function or program.</p>
<p>There are a number of ways to control program flow in R but the most common is an if/else statment so we’ll focus on that one.</p>
<div id="ifelse" class="section level4">
<h4>If/Else</h4>
<p>There are two main parts to an if statement. The part where we make the decision (if) and then the code following the if statement that is excuted when the if statment is true.</p>
<pre class="r"><code>cats = &quot;cool&quot;

if (cats == &quot;cool&quot;) {
    print(&quot;You know it!&quot;)
}
## [1] &quot;You know it!&quot;</code></pre>
<p>Note the syntax here: Double equals (==) to signify comparision instead of assignment and the use of curly brackets to indicate the code we run when the if statement is true.</p>
<p>Adding an else statement allows to define code to run when the if statement is false.</p>
<pre class="r"><code>cats = &quot;uncool&quot;

if (cats == &quot;cool&quot;) {
    print(&quot;You know it!&quot;)
} else {
    print(&quot;Whaaat?!&quot;)
}
## [1] &quot;Whaaat?!&quot;</code></pre>
<p>If statements can be nested:</p>
<pre class="r"><code>dogs = &quot;cool&quot;

if (cats == &quot;cool&quot;) {
    print(&quot;You know it!&quot;)
} else if (dogs == &quot;cool&quot; ) {
    print(&quot;Phew, I was worried about you there.&quot;)
} else {
    print(&quot;Whaaat?!&quot;)
}
## [1] &quot;Phew, I was worried about you there.&quot;</code></pre>
<p>This can be a good way to choose from a list options but be careful about nests, within nests, within nests.</p>
</div>
<div id="logical-operators" class="section level4">
<h4>Logical operators</h4>
<p>Any statement that returns a TRUE or FALSE can be used in an if statement.</p>
<pre class="r"><code>3 + 2 == 4
## [1] FALSE

any(is.na(x))
## [1] FALSE

some_values = c(1,2,3,4)
3 %in% some_values
## [1] TRUE

# The %in% function retuns TRUE or FALSE for every value on the left-hand side depending on if a
# match is found on the right hand side.  This can be useful for if statements but only if it
# returns a single TRUE/FALSE

test_values = c(0,3,5)
if (test_values %in% some_values) {
    print(&quot;Yup&quot;)
}
## Warning in if (test_values %in% some_values) {: the condition has length &gt;
## 1 and only the first element will be used

test_values %in% some_values
## [1] FALSE  TRUE FALSE</code></pre>
<p>Here are some common logical operators that can be used in an if statement.</p>
<table>
<thead>
<tr class="header">
<th>Operator</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>&lt;</td>
<td>less than</td>
</tr>
<tr class="even">
<td>&lt;=</td>
<td>less than or equal to</td>
</tr>
<tr class="odd">
<td>&gt;</td>
<td>greater than</td>
</tr>
<tr class="even">
<td>&gt;=</td>
<td>greather than or equal to</td>
</tr>
<tr class="odd">
<td>==</td>
<td>exactly equal to</td>
</tr>
<tr class="even">
<td>!=</td>
<td>not equal to</td>
</tr>
<tr class="odd">
<td>!x</td>
<td>not x</td>
</tr>
<tr class="even">
<td>x</td>
<td>y</td>
</tr>
<tr class="odd">
<td>x &amp; y</td>
<td>x AND y</td>
</tr>
<tr class="even">
<td>isTRUE(x)</td>
<td>test if x is TRUE</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="assignment" class="section level2">
<h2>Assignment</h2>
<p>A quick note here on two different assignment operators used in R. Historically R has used <code>&lt;-</code> for assignment.</p>
<pre class="r"><code>x &lt;- 5
x
## [1] 5</code></pre>
<p>However, in this course so far you’ve seen me using <code>=</code> for assignment.</p>
<pre class="r"><code>x = 5
x
## [1] 5</code></pre>
<p>Both are equally valid, despite what you may read otherwise. Each has a couple of quirks to be aware of but these are very minor. Decide which one you prefer and be consistent.</p>
<div id="using--" class="section level3">
<h3>Using <code>&lt;-</code></h3>
<ul>
<li>Longer to type (two key strokes, plus SHIFT)</li>
<li>Can make mistakes like this: FALSE</li>
<li>Your code will look more like the majority of what’s out there.</li>
</ul>
</div>
<div id="using" class="section level3">
<h3>Using <code>=</code></h3>
<ul>
<li>Quicker to type</li>
<li>Similar to modern programming lanugages</li>
<li>Also used for passing parameters in functions <code>rnorm(n = 10)</code></li>
<li>Difference between <code>=</code> and <code>==</code> can be confusing to start</li>
</ul>
<p>Decide for yourself, be consistent and whichever you choose make sure to surround it in spaces.</p>
<pre class="r"><code># Good
x = 5 
y &lt;- 4

# Bad
x=5
y&lt;-4</code></pre>
</div>
</div>
<div id="environments" class="section level2">
<h2>Environments</h2>
<p>What is an environment? This is a topic for an entire workshop in and of itself, however it is important to have a basic understanding of environments.</p>
<p>Environments are how R knows where to look for things. The only environment that you <em>usually</em> have to pay attention to is <code>.GlobalEnv</code>, your working environment. When you define an object <code>z = 50</code>, this object, <code>z</code> now lives in the global environment. When you ask R to do something with the object <code>z</code>, say <code>print(z)</code>, then R begins to look for object <code>z</code> in the global environment. If it can’t find it there then it searches other places it knows like attached packages. For example if <code>z</code> happened to be a function in a package called <code>alphabet</code> and that package had been attached (<code>library(alphabet)</code>), then R would find <code>z</code> there.</p>
<p>Environments are important to understand even when you are starting out because they can be the source of hard to find but devestating mistakes. <strong>Know what is in your environment!</strong> RStudio has a very useful panel called <code>Environment</code> that tells you exactly what is in your global environment. The function <code>ls()</code> also lists the objects in your global environment.</p>
<p>Here’s the kicker - you can define objects of any name in your global environment. Here’s something you should never do. Best not to run this code in your own session.</p>
<pre class="r"><code>5 + 5 
## [1] 10

`+` = function(x,y) {
    return(x*y)
}

5 + 5
## [1] 25

ls()
## [1] &quot;+&quot;           &quot;cats&quot;        &quot;dogs&quot;        &quot;say_hello&quot;   &quot;some_values&quot;
## [6] &quot;test_values&quot; &quot;x&quot;           &quot;y&quot;

rm(&quot;+&quot;)

5 + 5
## [1] 10</code></pre>
<p>Why does this work? Hint: Think about where R looks first to find objects.</p>
<p>Some tips:</p>
<ul>
<li>Always start your analysis in a new environment</li>
<li>Never save your workspace (R asks you about this when quitting, always say no)</li>
<li>Don’t analyze different projects in the same environment.</li>
<li>If you have an error try re-running your script in a new session (fresh environment)</li>
<li>On the flipside, make sure your script can run in a new session.</li>
<li>Use unique names for your objects</li>
</ul>
</div>
<div id="how-to-work-with-objects-subsetting-indexing-apply-functions" class="section level2">
<h2>How to work with objects: subsetting, indexing, apply functions</h2>
<div id="subsetting" class="section level3">
<h3>Subsetting</h3>
<p>Some common ways to subset dataframes</p>
<pre class="r"><code>head(iris)
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
## 3          4.7         3.2          1.3         0.2  setosa
## 4          4.6         3.1          1.5         0.2  setosa
## 5          5.0         3.6          1.4         0.2  setosa
## 6          5.4         3.9          1.7         0.4  setosa

# just the species column
head(iris$Species)
## [1] setosa setosa setosa setosa setosa setosa
## Levels: setosa versicolor virginica

# 3rd row and all columns
iris[3, ]
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 3          4.7         3.2          1.3         0.2  setosa

# 3rd column and all rows
head(iris[, 3])
## [1] 1.4 1.4 1.3 1.5 1.4 1.7

# 4th row and 2nd column
iris[4, 2]
## [1] 3.1

# rows 1 to 10 and all columns
iris[1:10, ]
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1           5.1         3.5          1.4         0.2  setosa
## 2           4.9         3.0          1.4         0.2  setosa
## 3           4.7         3.2          1.3         0.2  setosa
## 4           4.6         3.1          1.5         0.2  setosa
## 5           5.0         3.6          1.4         0.2  setosa
## 6           5.4         3.9          1.7         0.4  setosa
## 7           4.6         3.4          1.4         0.3  setosa
## 8           5.0         3.4          1.5         0.2  setosa
## 9           4.4         2.9          1.4         0.2  setosa
## 10          4.9         3.1          1.5         0.1  setosa</code></pre>
<p>Using vectors</p>
<pre class="r"><code># predfined vector
rows_to_keep = sample(1:nrow(iris), size = 10)
new_iris = iris[rows_to_keep, ]
new_iris
##     Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
## 61           5.0         2.0          3.5         1.0 versicolor
## 70           5.6         2.5          3.9         1.1 versicolor
## 39           4.4         3.0          1.3         0.2     setosa
## 31           4.8         3.1          1.6         0.2     setosa
## 114          5.7         2.5          5.0         2.0  virginica
## 137          6.3         3.4          5.6         2.4  virginica
## 105          6.5         3.0          5.8         2.2  virginica
## 149          6.2         3.4          5.4         2.3  virginica
## 147          6.3         2.5          5.0         1.9  virginica
## 98           6.2         2.9          4.3         1.3 versicolor

# logicals work too
new_iris$Species == &quot;setosa&quot;
##  [1] FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE

new_iris[new_iris$Species == &quot;setosa&quot;, ]
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 39          4.4         3.0          1.3         0.2  setosa
## 31          4.8         3.1          1.6         0.2  setosa</code></pre>
<p>For more details on subsetting read the <a href="http://adv-r.had.co.nz/Subsetting.html">subsetting</a> chapter in Advanced R</p>
</div>
</div>
<div id="loops-and-apply" class="section level2">
<h2>Loops and apply</h2>
<div id="loops" class="section level3">
<h3>Loops</h3>
<p>Many times when programming, a task, or set of tasks, need to be done over and over again on lots of things. For instance lets say that you have a data frame and want to calcluate the mean of each column.</p>
<pre class="r"><code>df = data.frame(x = rnorm(50), y = rnorm(50), z = rnorm(50))
head(df)
##            x          y           z
## 1 -1.0473162 -1.0170784  2.36413015
## 2 -1.3377041  0.6559729  0.08380648
## 3  1.3103186  1.0566558  0.79733153
## 4  0.9018513  1.5615109 -0.35769775
## 5  1.0071457  0.1213492  0.64683362
## 6  2.6144762 -1.6651174  0.63830821</code></pre>
<p>It could be done individually:</p>
<pre class="r"><code>mean(df$x)
## [1] 0.3098777
mean(df$y)
## [1] 0.05578127
mean(df$z)
## [1] 0.03511254</code></pre>
<p>But this gets tedious for many columns and it’s easy to make mistakes. Loops are a good choice here.</p>
<pre class="r"><code>output = vector(&quot;double&quot;, ncol(df))

for (i in seq_along(df)) {
    output[[i]] = mean(df[[i]])
}

output
## [1] 0.30987769 0.05578127 0.03511254</code></pre>
<div id="break-it-down" class="section level4">
<h4>Break it down:</h4>
<p>First predefine our output. Do this when you can as it speeds up loops in R significantly.</p>
<pre class="r"><code>output = vector(&quot;double&quot;, ncol(df))</code></pre>
<p>Next we need something to loop over, something that changes with each turn of the loop. <code>seq_along()</code> does the trick here, but this can really be anything</p>
<pre class="r"><code>seq_along(df)
## [1] 1 2 3

# equivalently
1:ncol(df)
## [1] 1 2 3

1:3
## [1] 1 2 3</code></pre>
<p>Now we have the variable <code>i</code> which changes each pass of the loop. We can use this to calculate the mean of each column using what we learned about subsetting.</p>
<pre class="r"><code>output[[i]] = mean(df[[i]])</code></pre>
<p>Bonus: Why do we have to use <code>[[</code> here?</p>
</div>
<div id="some-notes-on-loops" class="section level4">
<h4>Some notes on loops</h4>
<ul>
<li>The code that you want to be evaluated in the loop must be between the curly brackes <code>{}</code>.</li>
<li>A more general loop, the <code>while</code> loop can also be used but is not often seen in R code.</li>
<li>In R, looping over very large objects, for example a data frame with a million rows, can be very slow, especially if you don’t predefine your output.</li>
<li>For this reason, in many cases it is usually more efficient and often cleaner to use an <code>apply</code> function.</li>
</ul>
</div>
</div>
<div id="apply" class="section level3">
<h3>Apply</h3>
<p>The apply family of functions can be some of the most useful but most confusing functions to use. They are both fast and powerful but do remember that loops work too, despite what you might read in forums.</p>
<p>Here we’ll just a brief introduction so you are familiar with them, as they are topic more suitable for an advanced workshop.</p>
</div>
</div>
<div id="other-important-key-r-functions" class="section level2">
<h2>Other important key R functions</h2>
<div id="generating-and-manipulating-sequences" class="section level3">
<h3>Generating and manipulating sequences</h3>
<pre class="r"><code>seq(1,10)
##  [1]  1  2  3  4  5  6  7  8  9 10

seq(1, 20, by = 2)
##  [1]  1  3  5  7  9 11 13 15 17 19

seq(from = 2, by = 2.5, length.out = 10)
##  [1]  2.0  4.5  7.0  9.5 12.0 14.5 17.0 19.5 22.0 24.5

rep(2, 3)
## [1] 2 2 2

rep(1:3, 4)
##  [1] 1 2 3 1 2 3 1 2 3 1 2 3

rep(1:3, each = 2)
## [1] 1 1 2 2 3 3

rep(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), each = 6)
##  [1] &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;B&quot; &quot;B&quot; &quot;B&quot; &quot;B&quot; &quot;B&quot; &quot;B&quot; &quot;C&quot; &quot;C&quot; &quot;C&quot; &quot;C&quot; &quot;C&quot;
## [18] &quot;C&quot;</code></pre>
<pre class="r"><code>normal_dist = rnorm(n = 1e6, mean = 2, sd = 1.2)
hist(normal_dist, main = &quot;A normal distribution&quot;)</code></pre>
<p><img src="2_understanding_R_files/figure-html/unnamed-chunk-23-1.png" width="672" /></p>
<pre class="r"><code>
uniform_dist = runif(n = 1e6, min = 10, max = 20)
hist(uniform_dist, main = &quot;A uniform distribution&quot;)</code></pre>
<p><img src="2_understanding_R_files/figure-html/unnamed-chunk-23-2.png" width="672" /></p>
</div>
<div id="basic-statistics" class="section level3">
<h3>Basic statistics</h3>
<pre class="r"><code>set.seed(3823)

x = sample(1:1000, size = 50, replace = TRUE)

max(x)
## [1] 982

min(x)
## [1] 4

range(x)
## [1]   4 982

mean(x)
## [1] 511.7

median(x)
## [1] 511

sum(x)
## [1] 25585

sd(x)
## [1] 265.8911

y = rnorm(x, 1, 0.2 * x) + x
plot(x,y)</code></pre>
<p><img src="2_understanding_R_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p>
<pre class="r"><code>
var(x)
## [1] 70698.09

cor(x,y)
## [1] 0.9351317</code></pre>
<pre class="r"><code>my_model = lm(y ~ x)

print(my_model)
## 
## Call:
## lm(formula = y ~ x)
## 
## Coefficients:
## (Intercept)            x  
##     -36.125        1.121

summary(my_model)
## 
## Call:
## lm(formula = y ~ x)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -278.58  -50.71    9.29   53.68  326.62 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) -36.1253    35.2760  -1.024    0.311    
## x             1.1210     0.0613  18.286   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 114.1 on 48 degrees of freedom
## Multiple R-squared:  0.8745, Adjusted R-squared:  0.8719 
## F-statistic: 334.4 on 1 and 48 DF,  p-value: &lt; 2.2e-16

plot(x, y, main = &quot;Linear model&quot;)
abline(my_model, col = &quot;red&quot;)</code></pre>
<p><img src="2_understanding_R_files/figure-html/unnamed-chunk-25-1.png" width="672" /></p>
</div>
</div>
<div id="homework" class="section level2">
<h2>Homework</h2>
<p>Install the readr package and using the readr documentation figure out how to load the <code>pa_phenotype_data.csv</code> dataset. With this dataset answer the following:</p>
<ul>
<li>How many rows and columns are in the data?</li>
<li>What is the maximum value in the “LB” column?</li>
<li>Are variables “Protease” and “Gelatinase” linearly correlated?</li>
<li>What is the adjusted R-squared?</li>
<li>Write a function that mean centers a column (subtract the column mean from every value in that column) and run it on the last 6 columns in the dataset.</li>
</ul>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
