<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Practical Data Management</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/lumen.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="site_libs/highlight/default.css"
      type="text/css" />
<script src="site_libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 54px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 59px;
  margin-top: -59px;
}

.section h2 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h3 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h4 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h5 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h6 {
  padding-top: 59px;
  margin-top: -59px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.9em;
  padding-left: 5px;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
  padding-left: 10px;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">UCVM R Bootcamp</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="overview.html">
    <span class="fa fa-map-signs"></span>
     
    Overview
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Course Materials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="2_understanding_R.html">Understanding R</a>
    </li>
    <li>
      <a href="3_prac_data_management_1.html">Practical data management 1</a>
    </li>
    <li>
      <a href="4_prac_data_management_2.html">Practical data management 2</a>
    </li>
    <li>
      <a href="5_data_viz.html">Data visualization</a>
    </li>
    <li>
      <a href="6_intro_to_bioconductor.html">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="7_wrapup.html">Reproducability and wrapup</a>
    </li>
    <li>
      <a href="8_Homework_Answers.html">Homework Answers</a>
    </li>
  </ul>
</li>
<li>
  <a href="references.html">
    <span class="fa fa-external-link"></span>
     
    References
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    About
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="about.html">
        <span class="fa fa-info-circle"></span>
         
        Who we are
      </a>
    </li>
    <li>
      <a href="license-summary.html">
        <span class="fa fa-creative-commons"></span>
         
        License
      </a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ucvm/rbootcamp">
    <span class="fa fa-github fa-2x"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Practical Data Management</h1>

</div>


<pre class="r"><code>library(&quot;dplyr&quot;)
library(&quot;tidyr&quot;)
library(&quot;ggplot2&quot;)
library(&quot;readr&quot;)</code></pre>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>The data is from a genotype test where the sequence of part of the HIV pol gene is sequenced. The patient sequence is compared to a database to identify resistance causing mutations.</p>
<p>The 3 classes of drugs tested are NRTI, NNRTI, PI. Donâ€™t worry about drug specifics for the purposes of this exercise</p>
<p>For this dataset, no drug resistance for that class is coded as 0 and resistance is coded as 1.</p>
</div>
<div id="loading-the-data" class="section level2">
<h2>Loading the data</h2>
<p>Load database results for patients in study.</p>
<pre class="r"><code>db = read_csv(&quot;datasets/DB_results.csv&quot;)
## Parsed with column specification:
## cols(
##   PATID = col_integer(),
##   First_pos_test = col_date(format = &quot;&quot;),
##   ENUM = col_character(),
##   DRAWN_DATE = col_date(format = &quot;&quot;),
##   NRTI = col_integer(),
##   NNRTI = col_integer(),
##   PI = col_integer()
## )</code></pre>
<p>Load table with all the sequences for the clinic.</p>
<pre class="r"><code>seqs = read_csv(&quot;datasets/All_sequences.csv&quot;)
## Parsed with column specification:
## cols(
##   PATID = col_integer(),
##   DRAWN_DATE = col_date(format = &quot;&quot;),
##   LOCATION = col_character(),
##   ENUM = col_character(),
##   TESTCODE = col_character(),
##   STRING = col_character()
## )</code></pre>
</div>
<div id="looking-at-our-data" class="section level2">
<h2>Looking at our data</h2>
<p>What did we just load?!</p>
<pre class="r"><code>glimpse(db)
## Observations: 23
## Variables: 7
## $ PATID          &lt;int&gt; 4, 4, 4, 5, 5, 5, 5, 5, 5, 1, 1, 1, 7, 7, 7, 7,...
## $ First_pos_test &lt;date&gt; 1995-07-10, 1995-07-10, 1995-07-10, 1999-02-01...
## $ ENUM           &lt;chr&gt; &quot;104A&quot;, &quot;083A&quot;, &quot;274A&quot;, &quot;160A&quot;, &quot;266A&quot;, &quot;314A&quot;,...
## $ DRAWN_DATE     &lt;date&gt; 2001-10-23, 2008-10-15, 2011-09-22, 2003-09-04...
## $ NRTI           &lt;int&gt; 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1,...
## $ NNRTI          &lt;int&gt; 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1,...
## $ PI             &lt;int&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...
glimpse(seqs)
## Observations: 33
## Variables: 6
## $ PATID      &lt;int&gt; 12, 1, 1, 1, 13, 14, 15, 2, 3, 3, 3, 3, 3, 4, 4, 4,...
## $ DRAWN_DATE &lt;date&gt; 2003-09-22, 2006-02-06, 2006-12-19, 2010-05-25, 20...
## $ LOCATION   &lt;chr&gt; &quot;Calgary_AB&quot;, &quot;Calgary_AB&quot;, &quot;Calgary_AB&quot;, &quot;Calgary_...
## $ ENUM       &lt;chr&gt; &quot;180A&quot;, &quot;241A&quot;, &quot;562B&quot;, &quot;318A&quot;, &quot;292A&quot;, &quot;201A&quot;, &quot;02...
## $ TESTCODE   &lt;chr&gt; &quot;VIRCOGENPR&quot;, &quot;POL_1017&quot;, &quot;VIRCOGENPR&quot;, &quot;VIRCOGENPR...
## $ STRING     &lt;chr&gt; &quot;CCTCAAATCACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAGGGGG...</code></pre>
<p>Is this tidy data? Why or why not?</p>
</div>
<div id="preparing-our-data" class="section level2">
<h2>Preparing our data</h2>
<div id="joining-tables" class="section level3">
<h3>Joining tables</h3>
<p>To do our analysis we need to combined the information from the two tables. To do this weâ€™ll use a join operation. Here we use a left join which will add the columns from the sequences table to the db table only where there is a match base on a â€˜keyâ€™ column. Weâ€™ll let the <code>left_join()</code> function decide which columns to use as keys, based on column names</p>
<pre class="r"><code>all = left_join(db, seqs)
## Joining, by = c(&quot;PATID&quot;, &quot;ENUM&quot;, &quot;DRAWN_DATE&quot;)
all
## # A tibble: 23 Ã— 10
##    PATID First_pos_test  ENUM DRAWN_DATE  NRTI NNRTI    PI    LOCATION
##    &lt;int&gt;         &lt;date&gt; &lt;chr&gt;     &lt;date&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;       &lt;chr&gt;
## 1      4     1995-07-10  104A 2001-10-23     0     0     0  Calgary_AB
## 2      4     1995-07-10  083A 2008-10-15     0     0     0  Calgary_AB
## 3      4     1995-07-10  274A 2011-09-22     0     0     0  Calgary_AB
## 4      5     1999-02-01  160A 2003-09-04     1     1     0  Calgary_AB
## 5      5     1999-02-01  266A 2005-04-21     1     1     0  Calgary_AB
## 6      5     1999-02-01  314A 2008-08-26     0     1     0  Calgary_AB
## 7      5     1999-02-01  090B 2009-09-02     1     1     0 Edmonton_AB
## 8      5     1999-02-01  493A 2010-10-20     0     0     0  Calgary_AB
## 9      5     1999-02-01  897A 2015-11-17     0     1     0  Calgary_AB
## 10     1     1986-08-06  241A 2006-02-06     0     0     0  Calgary_AB
## # ... with 13 more rows, and 2 more variables: TESTCODE &lt;chr&gt;,
## #   STRING &lt;chr&gt;</code></pre>
</div>
<div id="clean-up-column-names" class="section level3">
<h3>Clean up column names</h3>
<pre class="r"><code>all = all %&gt;% rename(Pat_ID = PATID, 
                             First_positive_test = First_pos_test, 
                             Geno_ID = ENUM,
                             Geno_date = DRAWN_DATE,
                             Location = LOCATION,
                             Sequence = STRING)</code></pre>
<p>And we donâ€™t want the <code>TESTCODE</code> column, letâ€™s drop it</p>
<pre class="r"><code>all = all %&gt;% select(-TESTCODE)</code></pre>
</div>
<div id="tidy-it" class="section level3">
<h3>Tidy it</h3>
<p>Finally letâ€™s get our tidy data</p>
<pre class="r"><code>final = all %&gt;% gather(Drug, Resistance, NRTI, NNRTI, PI)</code></pre>
<p>Note that I didnâ€™t want intermediate copies of the data here so Iâ€™ve simply keep using the same object name to overwrite the previous one. Sometimes in your own analysis you may want to keep certain intermediate copies.</p>
</div>
</div>
<div id="analysis" class="section level2">
<h2>Analysis</h2>
<p>First we want to calculate the sum of the resistance for all three drugs for eacch genotype results</p>
<pre class="r"><code>meta = final %&gt;% 
    group_by(Geno_ID, Pat_ID) %&gt;% 
    summarise(sum_resistance = sum(Resistance))</code></pre>
<p>Currently, HIV is not cured by the ARV drugs, only suppressed in the body. Therefore patients can acumulate resistance over time. If you look at the data, you can see that some patients have different resistance results. This is because the PCR used in the genotype only records the most abundant HIV variant circulating at that time but there are many variants in the population of that patient so we often assume that resistance is cumulative and therefore we want to categorize patients by their maximum resistance score ever (i.e.Â cumulative resistance).</p>
<pre class="r"><code>cumm_res = meta %&gt;%
  group_by(Pat_ID) %&gt;%
  summarise(max_res = max(sum_resistance))</code></pre>
<pre class="r"><code>meta = left_join(meta, cumm_res, by = &quot;Pat_ID&quot;)</code></pre>
<p>Count the number of tests per patient in each category</p>
<pre class="r"><code>ccr_counts = meta %&gt;%
  group_by(Pat_ID, max_res) %&gt;%
  summarise(freq = length(max_res)) </code></pre>
<p>Plot the results in a boxplot</p>
<pre class="r"><code>ccr_counts %&gt;% 
    ggplot(aes(x = factor(max_res), y = freq)) +
        geom_boxplot() +
      theme_bw() +
      labs(x = &quot;Cumulative class resistance/patient&quot;,
             y = &quot;Frequency of tests&quot;,
             title = &quot;Number of genotypes/patient by cumulative class resistance&quot;)</code></pre>
<p><img src="3_prac_data_management_1_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
